### PR TITLE
def => explicit types, fixes #205

### LINKED ISSUE #205: def => explicit types
Goal is to replace all uses of `def` with explicit types:

    // Before
    def count = 4
    def method() {
        return "string"
    }

    // After
    int count = 4
    String method() {
        return "string"
    }

The Groovy style guide indicates that it's a choice of using `def` or explicit types:

    So make a choice, either use def or a type.

Though `def` is more of the idiom for Groovy, explicit types help with correctness, catching errors in the IDE as a def isn't as constrained:

    When using def in Groovy, the actual type holder is Object (so you can assign any
    object to variables defined with def, and return any kind of object if a method is
    declared returning def).

Thus, a def is less restrictive on types. According to the following link in the Style Guide, it can be thought of as an alias to `Object` and the programmer indicating that they don't care about the type: http://www.groovy-lang.org/semantics.html#_variable_definition

### PR DESCRIPTION
**Pull request by brunobowden**
- Figured out tricky explicit type for IncrementalTaskInputs
- Explicit types for J2objcNativeCompilation
- Explicit types for iterators

TESTED=yes

Verified:
a) `grep -R "def " .` => only shows Style Guidelines
b) Explicit types for all “->” statements
c) Explicit types for all implicit "it" iterators

### CODE REVIEW

**Comment by brunobowden** [comment]
@advayDev1, @confile - I think these are the last changes needed for adding explicit types. The one thing I haven't been able to track down are these iterators for the Native Build:

screen shot 2015-06-24 at 6 39 09 pm

**Comment by brunobowden** [comment]
@advayDev1, @confile - this should be the final commit for making all types explicit.

I've also created a "0.3 Release" Milestone. Please add relevant items to that.

File: src/main/groovy/com/github/j2objccontrib/j2objcgradle/tasks/J2objcTranslateTask.groovy [hunk 1]
```diff
@@ -122,30 +124,34 @@ class J2objcTranslateTask extends DefaultTask {
...
-            // TODO: figure out the explicit type for change
-            inputs.outOfDate { change ->
-                // We must filter by srcFiles, since all possible input files are @InputFiles to this task.
-                if (originalSrcFiles.contains(change.file)) {
-                    logger.debug "New or Updated file: " + change.file
-                    srcFilesChanged += project.files(change.file)
-                } else {
-                    nonSourceFileChanged = true
-                    logger.debug "New or Updated non-source file: " + change.file
+            inputs.outOfDate(new Action<InputFileDetails>() {
```
Comment (advayDev1) [inline src/main/groovy/com/github/j2objccontrib/j2objcgradle/tasks/J2objcTranslateTask.groovy]: hmm... why is this needed?  (the wrapper with new Action)?

**Comment by confile** [comment]
I cannot add anything I am not a contributor. 


Michael

File: src/main/groovy/com/github/j2objccontrib/j2objcgradle/J2objcNativeCompilation.groovy [hunk 2]
```diff
@@ -71,14 +71,14 @@ class J2objcNativeCompilation {
             file("${buildDir}/j2objcHackToForceCompilation/Empty.m").createNewFile()
             file("${buildDir}/j2objcHackToForceCompilation/EmptyTest.m").createNewFile()
 
-            def simulatorClangArgs = [
+            String[] simulatorClangArgs = [
```
Comment (confile) [inline src/main/groovy/com/github/j2objccontrib/j2objcgradle/J2objcNativeCompilation.groovy]: isn't that List<String>

File: src/main/groovy/com/github/j2objccontrib/j2objcgradle/J2objcNativeCompilation.groovy [hunk 3]
```diff
@@ -96,63 +96,83 @@ class J2objcNativeCompilation {
                     // https://docs.gradle.org/current/userguide/nativeBinaries.html#withArguments
                     clang(Clang) {
                         target('ios_arm64') {
-                            def iosClangArgs = [
+                            String[] iosClangArgs = [
```
Comment (confile) [inline src/main/groovy/com/github/j2objccontrib/j2objcgradle/J2objcNativeCompilation.groovy]: List<String>  ??

File: src/main/groovy/com/github/j2objccontrib/j2objcgradle/J2objcNativeCompilation.groovy [hunk 4]
```diff
@@ -96,63 +96,83 @@ class J2objcNativeCompilation {
...
                             linker.withArguments { List<String> args ->
-                                iosClangArgs.each { args << it }
+                                iosClangArgs.each { String arg ->
+                                    args << arg
+                                }
                             }
                         }
                         target('ios_armv7') {
-                            def iosClangArgs = [
+                            String[] iosClangArgs = [
```
Comment (confile) [inline src/main/groovy/com/github/j2objccontrib/j2objcgradle/J2objcNativeCompilation.groovy]: List<String> ??

File: src/main/groovy/com/github/j2objccontrib/j2objcgradle/J2objcNativeCompilation.groovy [hunk 5]
```diff
@@ -96,63 +96,83 @@ class J2objcNativeCompilation {
...
                             linker.withArguments { List<String> args ->
-                                iosClangArgs.each { args << it }
+                                iosClangArgs.each { String arg ->
+                                    args << arg
+                                }
                             }
                         }
                         target('ios_armv7s') {
-                            def iosClangArgs = [
+                            String[] iosClangArgs = [
```
Comment (confile) [inline src/main/groovy/com/github/j2objccontrib/j2objcgradle/J2objcNativeCompilation.groovy]: List<String> ??

File: src/main/groovy/com/github/j2objccontrib/j2objcgradle/J2objcNativeCompilation.groovy [hunk 6]
```diff
@@ -96,63 +96,83 @@ class J2objcNativeCompilation {
...
                             linker.withArguments { List<String> args ->
-                                iosClangArgs.each { args << it }
+                                iosClangArgs.each { String arg ->
+                                    args << arg
+                                }
                             }
                         }
                         target('ios_i386') {
-                            def iosClangArgs = [
+                            String[] iosClangArgs = [
```
Comment (confile) [inline src/main/groovy/com/github/j2objccontrib/j2objcgradle/J2objcNativeCompilation.groovy]: List<String> ??

File: src/main/groovy/com/github/j2objccontrib/j2objcgradle/J2objcNativeCompilation.groovy [hunk 7]
```diff
@@ -96,63 +96,83 @@ class J2objcNativeCompilation {
...
                             linker.withArguments { List<String> args ->
-                                iosClangArgs.each { args << it }
+                                iosClangArgs.each { String arg ->
+                                    args << arg
+                                }
                             }
                         }
                         target('ios_x86_64') {
-                            def iosClangArgs = [
+                            String[] iosClangArgs = [
```
Comment (confile) [inline src/main/groovy/com/github/j2objccontrib/j2objcgradle/J2objcNativeCompilation.groovy]: List<String> ??

Comment (brunobowden) [inline src/main/groovy/com/github/j2objccontrib/j2objcgradle/J2objcNativeCompilation.groovy]: @confile - if there's 2 or 3 repeats, then feel free to mark it. If there's
more, just note it once and then mark (repeated below). I'll make sure I
modify all of them.

File: src/main/groovy/com/github/j2objccontrib/j2objcgradle/tasks/J2objcTranslateTask.groovy (hunk 1 above)
Comment (brunobowden) [inline src/main/groovy/com/github/j2objccontrib/j2objcgradle/tasks/J2objcTranslateTask.groovy]: I found it in a Gradle unit test:
https://github.com/gradle/gradle/blob/478a891a864c8749c194d42c52973cb1c8dfa8a7/subprojects/core/src/test/groovy/org/gradle/api/internal/changedetection/changes/DefaultTaskArtifactStateRepositoryTest.groovy#L558

Comment (brunobowden) [inline src/main/groovy/com/github/j2objccontrib/j2objcgradle/tasks/J2objcTranslateTask.groovy]: I can't seem to make all the type checking work without this. Since it's in an official Gradle unit test, I'm inclined to believe it's correct.

File: src/main/groovy/com/github/j2objccontrib/j2objcgradle/J2objcNativeCompilation.groovy (hunk 7 above)
Comment (brunobowden) [inline src/main/groovy/com/github/j2objccontrib/j2objcgradle/J2objcNativeCompilation.groovy]: @confile - another thoughts is to say `Fix XXX, repeat 7 times below`. That succinctly explain it and I'll find the other references.

File: src/main/groovy/com/github/j2objccontrib/j2objcgradle/tasks/J2objcTranslateTask.groovy (hunk 1 above)
Comment (advayDev1) [inline src/main/groovy/com/github/j2objccontrib/j2objcgradle/tasks/J2objcTranslateTask.groovy]: weird, we don't do that for other Action params that we pass as closures, but ok.

**Comment by advayDev1** [comment]
LGTM with above comments

**Comment by brunobowden** [comment]
The `String[]'s` are ok? I don't think that they're meant to be `List's`.

**Comment by advayDev1** [comment]
@brunobowden - yes, totally fine:
http://docs.groovy-lang.org/docs/latest/html/documentation/#_array_initializers

**Comment by advayDev1** [comment]
@brunobowden - sorry, i should have been clearer before.  it was LGTM with (my) above comments.  I was not saying you need to do @confile 's.

**Comment by brunobowden** [comment]
I understood it that way

File: src/main/groovy/com/github/j2objccontrib/j2objcgradle/J2objcNativeCompilation.groovy (hunk 7 above)
Comment (confile) [inline src/main/groovy/com/github/j2objccontrib/j2objcgradle/J2objcNativeCompilation.groovy]: @brunobowden Okay I will do that. But what about the fix?

Comment (brunobowden) [inline src/main/groovy/com/github/j2objccontrib/j2objcgradle/J2objcNativeCompilation.groovy]: Hi @confile, @advayDev1 confirmed that doing the `String[]` initialization was ok:

http://docs.groovy-lang.org/docs/latest/html/documentation/#_array_initializers

See previous comment from him:

screen shot 2015-06-26 at 7 00 16 am

Comment (confile) [inline src/main/groovy/com/github/j2objccontrib/j2objcgradle/J2objcNativeCompilation.groovy]: ok fine

**Comment by confile** [comment]
LGTM
